engines/e_pkcs11_eng.c

```diff
@@ -0,0 +1,598 @@
+        i = EVP_read_pw_string((char *)strbuf, buflen, prompt, 1);
+        if (i == 0) {
+            if (strbuf[0] == '\0') {
+                goto err;
+            }
+            *pin = strbuf;
+            return 1;
+        }
+        if (i < 0) {
+            PKCS11_trace("bad password read\n");
+            goto err;
+        }
+    }
+
+ err:
+    OPENSSL_free(strbuf);
+#endif
+
+    return 0;
+}
+
+static int pkcs11_parse(PKCS11_CTX *ctx, const char *path)
+{
+    char *id, *module_path = NULL;
+    char *pin = NULL, *label = NULL, *slotid = NULL;
+
+    ctx->slotid = 0;
+    if (strncmp(path, "pkcs11:", 7) == 0) {
+        path += 7;
+        if (ctx->module_path == NULL &&
+            !pkcs11_parse_uri(path,"module-path=", &module_path))
+            goto err;
+        if (ctx->module_path == NULL) ctx->module_path = module_path;
+        if (!pkcs11_parse_uri(path,"id=", &id) &&
+            !pkcs11_parse_uri(path,"object=", &label)) {
+            PKCS11_trace("ID and OBJECT are null\n");
+            goto err;
+        }
+        if (!pkcs11_parse_uri(path,"slot-id=", &slotid)) {
+            slotid = NULL;
+        }
+        pkcs11_parse_uri(path,"pin-value=", &pin);
+        if (pin != NULL)
+            ctx->pin = (CK_BYTE *) pin;
+    } else if (path == NULL) {
+       PKCS11_trace("inkey is null\n");
+       goto err;
+    } else {
+        if (ctx->module_path == NULL) {
+            PKCS11_trace("Module path is null\n");
+            goto err;
+        }
+        id = OPENSSL_strdup(path);
+
+        if (id == NULL) {
+            PKCS11err(PKCS11_F_PKCS11_PARSE, ERR_R_MALLOC_FAILURE);
+            goto err;
+        }
+        if (pkcs11_ishex(id))
+            id = pkcs11_hex2a(id);
+
+        slotid = NULL;
+    }
+
+    if (label != NULL)
+        ctx->label = (CK_BYTE *) label;
+    else
+        ctx->id = (CK_BYTE *) id;
+
+    if (slotid != NULL)
+        ctx->slotid = (CK_SLOT_ID) atoi(slotid);
+
+    if (ctx->pin == NULL) {
+        if (!pkcs11_get_console_pin(&pin))
+            goto err;
+        ctx->pin = (CK_BYTE *) pin;
+        if (ctx->pin == NULL) {
+            PKCS11_trace("PIN is invalid\n");
+            goto err;
+        }
+    }
+    return 1;
+
+ err:
+    return 0;
+}
+
+static EVP_PKEY *pkcs11_engine_load_private_key(ENGINE * e, const char *path,
+                                                UI_METHOD * ui_method,
+                                                void *callback_data)
+{
+    CK_RV rv;
+    PKCS11_CTX *ctx;
+
+    ctx = ENGINE_get_ex_data(e, pkcs11_idx);
+    if (!pkcs11_parse(ctx, path))
+        goto err;
+
+    rv = pkcs11_initialize(ctx->module_path);
+
+    if (rv != CKR_OK)
+        goto err;
+    if (!pkcs11_get_slot(ctx))
+        goto err;
+    if (!pkcs11_start_session(ctx))
+        goto err;
+    if (!pkcs11_login(ctx, CKU_USER))
+        goto err;
+    if (!pkcs11_find_private_key(ctx))
+        goto err;
+
+    return pkcs11_load_pkey(ctx);
```

Comment:
As far as I can see, `pkcs11_load_pkey` only loads the public key components.  I haven't quite been able to see ...  oh wait I do, the rest of the key is found through the pkcs11 context...

This makes it difficult to do operations with multiple keys through this same engine, doesn't it?  What you really should do is to store a reference to the CK_OBJECT_HANDLE with the EVP_PKEY you return from `pkcs11_load_pkey`.  It's actually possible to do, although indirectly.  The following three lines from the CHIL engine over in [engine-corner](https://github.com/engine-corner) might give you a clue:

https://github.com/engine-corner/engine-chil/blob/493de783556bcf4e61e80f7f442babfb71d7887f/e_chil.c#L976-L978